gas-estimator/src/estimator.rs

```diff
@@ -0,0 +1,1222 @@
+        .asset_hub_client
+        .runtime_api()
+        .at_latest()
+        .await
+        .map_err(|e| {
+            EstimatorError::InvalidCommand(format!("Failed to get latest block: {:?}", e))
+        })?
+        .call(runtime_api_call)
+        .await
+        .map_err(|e| EstimatorError::InvalidCommand(format!("Failed to dry run XCM: {:?}", e)))?;
+
+    match dry_run_result {
+        Ok(effects) => {
+            let success = matches!(effects.execution_result, Outcome::Complete { .. });
+
+            let error_message = if success {
+                None
+            } else {
+                Some(format!(
+                    "XCM execution failed: {:?}",
+                    effects.execution_result
+                ))
+            };
+
+            let forwarded_xcms = effects.forwarded_xcms;
+
+            let forwarded_xcm = if forwarded_xcms.is_empty() {
+                None
+            } else {
+                Some(forwarded_xcms[0].clone())
+            };
+
+            Ok(DryRunResult {
+                success,
+                error_message,
+                forwarded_xcm,
+            })
+        }
+        Err(e) => Ok(DryRunResult {
+            success: false,
+            error_message: Some(format!("Dry run API error: {:?}", e)),
+            forwarded_xcm: None,
+        }),
+    }
+}
+
+#[cfg(feature = "local")]
+async fn dry_run_xcm_on_destination(
+    clients: &Clients,
+    forwarded_xcm: (VersionedLocation, Vec<VersionedXcm>),
+) -> Result<DryRunResult, EstimatorError> {
+    let (destination_location, xcms) = forwarded_xcm;
+
+    let para_id = extract_parachain_id(&destination_location)?;
+
+    let xcm = xcms
+        .into_iter()
+        .next()
+        .ok_or_else(|| EstimatorError::InvalidCommand("No XCM to forward".to_string()))?;
+
+    match para_id {
+        PENPAL_PARA_ID => penpal::dry_run_xcm(clients, xcm).await,
+        _ => Err(EstimatorError::InvalidCommand(format!(
+            "Unsupported destination parachain for local network: {}",
+            para_id
+        ))),
+    }
+}
+
+#[cfg(feature = "local")]
+async fn calculate_destination_execution_fee(
+    clients: &Clients,
+    forwarded_xcm: &(VersionedLocation, Vec<VersionedXcm>),
+    para_id: u32,
+    dot_asset: &Location,
+    ether_asset: &Location,
+) -> Result<(u128, u128), EstimatorError> {
+    let (_, xcms) = forwarded_xcm;
+    let xcm = xcms
+        .iter()
+        .next()
+        .ok_or_else(|| EstimatorError::InvalidCommand("No XCM to calculate fee for".to_string()))?;
+
+    match para_id {
+        PENPAL_PARA_ID => {
+            let fee_dot = crate::penpal::calculate_execution_fee(clients, xcm).await?;
+            let fee_ether = quote_price_exact_tokens_for_tokens(
+                clients,
+                dot_asset.clone(),
+                ether_asset.clone(),
+                fee_dot,
+                true,
+            )
+            .await?;
+
+            Ok((fee_dot, fee_ether))
+        }
```

Comment:
I’d suggest not using another dry run to estimate the cost on the destination side. IMHO, maintaining different parachain runtimes adds complexity.

To keep things simple for now, could we just maintain a map with the fee for each parachain instead? It could potentially come from the @snowbridge/registry package which I assume won't change that often.

Reply:
Yeah, the dry-running is definitely useful. But I assume we will be using dry-running in the Web-SDK anyway? 

In the case of transfers, I assume the XCM should follow a fixed format based on the transfer type (PNA or ENA), so the fee can be calculated via dry-run in advance (e.g. in @snowbridge/registry), then be reused here.

Or is there any way that this binary can be reused on web?

Reply:
@yrong I would also prefer not to duplicate the logic. My concern with using the SDK was that we shouldn't need to host an API for relayers to call. If we can package it as a binary i think this is an option.